File: app/frontend/src/lib/api/document-revisions.ts
import { apiFetch } from "../api-client";
import { mutationHeaders } from "../idempotency";

export interface DraftPageWrite {
  text: string;
  parent_revision_id: string;
  edit_reason?: string;
}

export interface DraftPageRead {
  page_revision_id: string;
  lock_version: number;
  page_number: number;
  text: string;
  status: string;
}

export interface RevisionSetRead {
  revision_set_id: string;
  document_id: string;
  revision_number: number;
  status: string;
  created_by_user_id: string;
  created_at: string | null;
  submitted_at: string | null;
  approved_by_user_id: string | null;
  approved_at: string | null;
}

export interface ApproveRevisionRequest {
  demo_mode?: boolean;
}

export interface RejectRevisionRequest {
  reason?: string;
}

export interface RestoreRevisionRequest {
  revision_id: string;
  reason?: string;
}

export interface GenerationAcceptedRead {
  generation_id: string;
  state: string;
}

export async function saveDraftPage(
  documentId: string,
  pageNumber: number,
  payload: DraftPageWrite,
  options: { idempotencyKey: string; lockVersion?: number },
): Promise<DraftPageRead> {
  return apiFetch<DraftPageRead>(`/documents/${documentId}/draft/pages/${pageNumber}`, {
    method: "PATCH",
    headers: mutationHeaders(options),
    body: JSON.stringify(payload),
  });
}

export async function submitDraft(
  documentId: string,
  options: { idempotencyKey: string; lockVersion?: number },
): Promise<RevisionSetRead> {
  return apiFetch<RevisionSetRead>(`/documents/${documentId}/draft/submit`, {
    method: "POST",
    headers: mutationHeaders(options),
  });
}

export async function approveRevisionSet(
  documentId: string,
  revisionSetId: string,
  payload: ApproveRevisionRequest,
  options: { idempotencyKey: string; lockVersion?: number },
): Promise<GenerationAcceptedRead> {
  return apiFetch<GenerationAcceptedRead>(
    `/documents/${documentId}/revision-sets/${revisionSetId}/approve`,
    {
      method: "POST",
      headers: mutationHeaders(options),
      body: JSON.stringify(payload),
    },
  );
}

export async function rejectRevisionSet(
  documentId: string,
  revisionSetId: string,
  payload: RejectRevisionRequest,
  options: { idempotencyKey: string; lockVersion?: number },
): Promise<RevisionSetRead> {
  return apiFetch<RevisionSetRead>(
    `/documents/${documentId}/revision-sets/${revisionSetId}/reject`,
    {
      method: "POST",
      headers: mutationHeaders(options),
      body: JSON.stringify(payload),
    },
  );
}

export async function restoreRevision(
  documentId: string,
  revisionSetId: string,
  payload: RestoreRevisionRequest,
  options: { idempotencyKey: string; lockVersion?: number },
): Promise<DraftPageRead> {
  return apiFetch<DraftPageRead>(
    `/documents/${documentId}/revision-sets/${revisionSetId}/restore`,
    {
      method: "POST",
      headers: mutationHeaders(options),
      body: JSON.stringify(payload),
    },
  );
}

export async function listRevisionSets(documentId: string): Promise<RevisionSetRead[]> {
  return apiFetch<RevisionSetRead[]>(`/documents/${documentId}/revision-sets`);
}

File: app/backend/src/hospital_ai/schemas/document_revisions.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class DraftPageWrite(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    text: str
    parent_revision_id: uuid.UUID
    edit_reason: str = ""


class DraftPageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    page_revision_id: uuid.UUID
    lock_version: int
    page_number: int
    text: str
    status: str


class RevisionSetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    revision_set_id: uuid.UUID
    document_id: uuid.UUID
    revision_number: int
    status: str
    created_by_user_id: uuid.UUID
    created_at: Optional[datetime] = None
    submitted_at: Optional[datetime] = None
    approved_by_user_id: Optional[uuid.UUID] = None
    approved_at: Optional[datetime] = None


class ApproveRevisionRequest(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    demo_mode: bool = False


class RejectRevisionRequest(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    reason: str = ""


class RestoreRevisionRequest(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    revision_id: uuid.UUID
    reason: str = ""


class GenerationAcceptedRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    generation_id: uuid.UUID
    state: str
